File: 186_connectedness_of_a_network/main.py
S = [0] * 10000000
for k in range(1, 56):
    S[k] = (100003 - 200003*k + 300007*k*k*k) % 1000000

for k in range(56, 10000000):
    S[k] = (S[k-24] + S[k-55]) % 1000000

n = 1
def get_next_call():
    global n
    caller = S[2*n - 1]
    called = S[2*n]
    n += 1
    return (caller, called)

def solution(NUMBER, p):
    people = [0] * 1000000
    groups = [[] for _ in range(1000000)]
    call = get_next_call()
    calls = 1
    group = 1
    prime_minister_friends = 0
    minister_called = False
    while (prime_minister_friends < p * 10000):
        if people[call[0]] == 0 and people[call[1]] == 0:
            people[call[0]] = group
            people[call[1]] = group
            groups[group].append(call[0])
            groups[group].append(call[1])
            group += 1
        elif people[call[0]] == 0:
            people[call[0]] = people[call[1]]
            groups[people[call[1]]].append(call[0])
        elif people[call[1]] == 0:
            people[call[1]] = people[call[0]]
            groups[people[call[0]]].append(call[1])
        elif people[call[0]] != people[call[1]]:
            group_to_delete = people[call[1]]
            for person in groups[group_to_delete]:
                people[person] = people[call[0]]
            groups[people[call[0]]] += groups[group_to_delete]
            groups[group_to_delete].clear()

        if not minister_called and (call[0] == NUMBER or call[1] == NUMBER):
            minister_called = True
        
        if minister_called:
            prime_minister_friends = len(groups[people[NUMBER]])

        call = get_next_call()
        while call[0] == call[1]:
            call = get_next_call()
        calls += 1
        
    return calls

print(solution(0, 1))

File: 186_connectedness_of_a_network/main.cpp
#include <iostream>
#include <vector>

void initialize_S(std::vector<int> &S) {
  S.push_back(0);
  long a, b, c;
  a = 100003;

  for (long i = 1; i <= 55; i++) {
    b = 200003 * i;
    c = 300007 * i * i * i;
    S.push_back((a - b + c) % 1000000);
  }
}

std::pair<int, int> get_next_call(std::vector<int> &S) {
  int len = S.size();
  S.push_back((S[len - 24] + S[len - 55]) % 1000000);
  len++;
  S.push_back((S[len - 24] + S[len - 55]) % 1000000);

  static int n = 1;
  int caller = S[2 * n - 1];
  int called = S[2 * n];
  n++;

  return std::pair<int, int>(caller, called);
}

int solution(const int NUMBER, const int p) {
  std::vector<int> S;
  initialize_S(S);

  std::vector<int> people(1000000, 0);
  std::vector<std::vector<int>> groups(1000000, std::vector<int>());
  int group = 1;
  auto call = get_next_call(S);
  int calls = 1;
  int prime_minister_friends = 0;
  bool minister_called = false;
  while (prime_minister_friends < p * 10000) {
    if (people[call.first] == 0 and people[call.second] == 0) {
      people[call.first] = group;
      people[call.second] = group;
      groups[group].push_back(call.first);
      groups[group].push_back(call.second);
      group++;
    } else if (people[call.first] == 0) {
      people[call.first] = people[call.second];
      groups[people[call.second]].push_back(call.first);
    } else if (people[call.second] == 0) {
      people[call.second] = people[call.first];
      groups[people[call.first]].push_back(call.second);
    } else if (people[call.first] != people[call.second]) {
      int group_second = people[call.second];
      std::vector<int> people_in_second_group = groups[group_second];
      for (int person : people_in_second_group) {
        people[person] = people[call.first];
        groups[people[call.first]].push_back(person);
      }
      groups[group_second].clear();
    }

    if (!minister_called && (call.first == NUMBER || call.second == NUMBER)) {
      minister_called = true;
    }
    if (minister_called) {
      if (groups[people[NUMBER]].size() > 10) {
        auto a = people[NUMBER];
        auto b = groups[a];
      }
      prime_minister_friends = groups[people[NUMBER]].size();
    }
    call = get_next_call(S);
    while (call.first == call.second) {
      call = get_next_call(S);
    }
    calls++;
  }
  return calls;
}

int main() {
  int NUMBER, p;
  std::cin >> NUMBER >> p;
  std::cout << solution(NUMBER, p);
}
